File: family-tree-frontend/src/services/familyTreeService.ts
import api from './api';
import { API_ENDPOINTS } from '../constants/apiEndpoints';
import {
  Person,
  PersonRequest,
  PersonPatchRequest,
  ApiResponse,
} from '../models/Person';

/**
 * Family Tree API Service
 * All API calls to the backend
 */

export const familyTreeService = {
  /**
   * Get the complete family tree
   */
  getFullTree: async (): Promise<ApiResponse<Person>> => {
    const response = await api.get<ApiResponse<Person>>(API_ENDPOINTS.FAMILY_TREE);
    return response.data;
  },

  /**
   * Get person by ID
   */
  getPersonById: async (id: string): Promise<ApiResponse<Person>> => {
    const response = await api.get<ApiResponse<Person>>(
      API_ENDPOINTS.FAMILY_TREE_BY_ID(id)
    );
    return response.data;
  },

  /**
   * Get person with all descendants
   */
  getPersonWithDescendants: async (id: string): Promise<ApiResponse<Person>> => {
    const response = await api.get<ApiResponse<Person>>(
      API_ENDPOINTS.FAMILY_TREE_DESCENDANTS(id)
    );
    return response.data;
  },

  /**
   * Create a new person
   */
  createPerson: async (data: PersonRequest): Promise<ApiResponse<Person>> => {
    const response = await api.post<ApiResponse<Person>>(
      API_ENDPOINTS.FAMILY_TREE,
      data
    );
    return response.data;
  },

  /**
   * Update person (PATCH)
   * Used for position updates and other partial updates
   */
  updatePerson: async (
    id: string,
    data: PersonPatchRequest
  ): Promise<ApiResponse<Person>> => {
    const response = await api.patch<ApiResponse<Person>>(
      API_ENDPOINTS.FAMILY_TREE_BY_ID(id),
      data
    );
    return response.data;
  },

  /**
   * Delete person
   */
  deletePerson: async (id: string): Promise<ApiResponse<void>> => {
    const response = await api.delete<ApiResponse<void>>(
      API_ENDPOINTS.FAMILY_TREE_BY_ID(id)
    );
    return response.data;
  },

  /**
   * Search persons by name
   */
  searchByName: async (name: string): Promise<ApiResponse<Person[]>> => {
    const response = await api.get<ApiResponse<Person[]>>(
      API_ENDPOINTS.FAMILY_TREE_SEARCH,
      { params: { name } }
    );
    return response.data;
  },

  /**
   * Get persons by generation level
   */
  getPersonsByLevel: async (level: number): Promise<ApiResponse<Person[]>> => {
    const response = await api.get<ApiResponse<Person[]>>(
      API_ENDPOINTS.FAMILY_TREE_BY_LEVEL(level)
    );
    return response.data;
  },

  /**
   * Get total person count
   */
  getTotalCount: async (): Promise<ApiResponse<number>> => {
    const response = await api.get<ApiResponse<number>>(
      API_ENDPOINTS.FAMILY_TREE_COUNT
    );
    return response.data;
  },

  /**
   * Reload data from JSON file (admin operation)
   */
  reloadData: async (): Promise<ApiResponse<void>> => {
    const response = await api.post<ApiResponse<void>>(
      API_ENDPOINTS.FAMILY_TREE_RELOAD
    );
    return response.data;
  },
};

File: api/src/main/java/com/familytree/service/FamilyTreeService.java
package com.familytree.service;

import com.familytree.dto.PersonDetailsRequest;
import com.familytree.dto.PersonDetailsResponse;
import com.familytree.dto.PersonPatchRequest;
import com.familytree.dto.PersonRequest;
import com.familytree.dto.PersonResponse;

import java.util.List;
import java.util.Optional;

/**
 * Service interface for family tree operations
 */
public interface FamilyTreeService {

	/**
	 * Get the complete family tree starting from root
	 */
	PersonResponse getFullTree();

	/**
	 * Get a specific person by ID with immediate children
	 */
	PersonResponse getPersonById(String id);

	/**
	 * Get a person by ID with full subtree
	 */
	PersonResponse getPersonWithAllDescendants(String id);

	/**
	 * Create a new person
	 */
	PersonResponse createPerson(PersonRequest request);

	/**
	 * Update an existing person (PATCH operation)
	 */
	PersonResponse updatePerson(String id, PersonPatchRequest request);

	/**
	 * Delete a person by ID
	 */
	void deletePerson(String id);

	/**
	 * Search persons by name
	 */
	List<PersonResponse> searchByName(String name);

	/**
	 * Get all persons at a specific generation level
	 */
	List<PersonResponse> getPersonsByLevel(Integer level);

	/**
	 * Get total count of persons in the tree
	 */
	long getTotalCount();

	/**
	 * Reload data from JSON file (admin operation)
	 */
	void reloadData();

	/**
	 * Reset all position values (positionX and positionY) to null
	 * This forces automatic layout recalculation on the frontend
	 */
	void resetAllPositions();

	/**
	 * Add or update person details
	 *
	 * @param personId the ID of the person
	 * @param request the details to add or update
	 * @return the created or updated details
	 */
	PersonDetailsResponse addOrUpdatePersonDetails(String personId, PersonDetailsRequest request);

	/**
	 * Get person details by person ID
	 *
	 * @param personId the ID of the person
	 * @return Optional containing the details if found
	 */
	Optional<PersonDetailsResponse> getPersonDetails(String personId);

	/**
	 * Delete person details
	 *
	 * @param personId the ID of the person
	 */
	void deletePersonDetails(String personId);
}
